[while_loop] support closures

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123018
* #123217

We add an additional_inputs arguments to the HOP while_loop and rename the operands to carried_inputs based on offline discussion with @zou3519 . This allows us to support closures, parameters and buffers.

The alternative is to pass the lifted inputs directly to outputs of body_fn. But since we want the body_fn's output to not aliasing input. We'll need to copy the inputs and remove the copies later. This is a bit more work to do. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

ydwu4 committed:
[while_loop] support closures

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] support closures"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

aakhundov commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -782,13 +781,22 @@ def _check_supported_callable(fn_var):
+                f"Expected a tuple but got {args[3].python_type()}",
+            )
+        additional_inputs = args[3].unpack_var_sequence(tx)
+        assert (
+            len(additional_inputs) == 0
+        ), "Additional inputs are set automatically by dynamo"
```
If additional inputs are set automatically by dynamo, what is the reason of passing an empty tuple here? Should the number of arguments match?

aakhundov commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -21,31 +21,40 @@
+            for t in (*carried_inputs, *additional_inputs)
+        ):
+            raise RuntimeError(
+                "carried_inputs must be a tuple of tensors, ints, floats, or bools, got "
```
`t` here can be from `additional_inputs`, too?

aakhundov commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -21,31 +21,40 @@
-        return super().__call__(cond_fn, body_fn, operands)
+        if not isinstance(additional_inputs, tuple):
+            raise RuntimeError(
+                "additional_inputs must be a tuple, got " f"{type(additional_inputs)}"
```
UX question: if `additional_inputs` are generated internally (and are *not* considered a part of public API), would the errors re. `additional_inputs` make sense to the user?

aakhundov commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -21,31 +21,40 @@
             raise RuntimeError(
-                "operands must be a tuple of tensors, ints, floats, or bools, got "
-                f"{operands}"
+                "carried_inputs must be a tuple, got " f"{type(carried_inputs)}"
```
Nit: why separate f-string here? And below.

aakhundov commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -97,31 +106,31 @@ def body_fn(iter, x):
     """
     if torch.compiler.is_dynamo_compiling():
-        return while_loop_op(cond_fn, body_fn, operands)
+        return while_loop_op(cond_fn, body_fn, carried_inputs, tuple())
```
Maybe add a comment here on what is the `tuple()` and why?

aakhundov commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -97,31 +106,31 @@ def body_fn(iter, x):
-def while_loop_dense(cond_fn, body_fn, operands):
-    init_val = operands
+def while_loop_dense(cond_fn, body_fn, carried_inputs, additional_inputs):
+    init_val = carried_inputs
```
Nit: by the end of this function, `init_val` are likely no longer initial values. Maybe rename to smth like `carried_vals`?

aakhundov commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -216,17 +239,22 @@ def while_loop_func(ctx, cond_fn, body_fn, operands):
+        ret = while_loop_op(
+            functional_cond_fn,
+            functional_body_fn,
+            unwrapped_carried_inputs,
+            unwrapped_additional_inputs,
```
Do we allow non-Tensor inputs in any of those?

ydwu4 replied:
Sorry, i was wrong. We've allowed tensor, and python int, float, bool to be passed in as inputs. Maybe we should add more tests for them.

aakhundov replied:
Inductor support for `torch.while_loop` currently only covers Tensor inputs (carried and additional), as we fetch fake tensors for them and pass those to subgraph lowering. We'll need special treatment of non-Tensor inputs.

ydwu4 replied:
We could defer it until a valid test cases I think.

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1544,21 +1544,30 @@ def codegen_conditional(self, conditional):
     def codegen_while_loop(self, while_loop):
         name = while_loop.get_name()
-        outer_inputs = [buf.codegen_reference() for buf in while_loop.operands]
+        outer_carried_inputs = [
+            buf.codegen_reference() for buf in while_loop.carried_inputs
+        ]
+        outer_additional_inputs = [
+            buf.codegen_reference() for buf in while_loop.additional_inputs
+        ]
+        outer_inputs = outer_carried_inputs + outer_additional_inputs
```
Nit: could as well combine `while_loop.carried_inputs + while_loop.additional_inputs` in the pre-existing list comprehension, as the same is done in both cases.

aakhundov replied:
Sure. Thanks for pointing out.

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1544,21 +1544,30 @@ def codegen_conditional(self, conditional):
+        outer_inputs = outer_carried_inputs + outer_additional_inputs
 
-        self.writeline(f"{name} = [None] * {len(while_loop.operands)}")
+        outer_input_len = len(while_loop.carried_inputs) + len(
+            while_loop.additional_inputs
+        )
```
Can use `len(outer_inputs)` downstream instead of `outer_input_len`?

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1544,21 +1544,30 @@ def codegen_conditional(self, conditional):
-            body_outer_inputs
-        )  # carry over the state from body_fn
+
+        # carry over the state from body_fn
```
Could we make this comment a bit more specific mentioning that we only carry over the `carried_inputs` part of the inputs, but not the additional ones?

aakhundov commented on torch/_inductor/ir.py:
```diff
@@ -7235,24 +7242,25 @@ def create(
 
         assert (
-            len(operands) > 0
+            len(carried_inputs) > 0
```
Actually, we can fetch the device from additional_inputs, too. So maybe just check that we have any input at least once?

aakhundov commented on torch/_inductor/ir.py:
```diff
@@ -7202,23 +7205,27 @@ def create(
 
-        fx_operands = V.graph.current_node.args[-1]
-        fake_operands = [x.meta["val"] for x in fx_operands]  # type: ignore[union-attr]
+        fx_carried_inputs = V.graph.current_node.args[-2]
+        fx_additional_inputs = V.graph.current_node.args[-1]
+        fake_carried_inputs = [x.meta["val"] for x in fx_carried_inputs]  # type: ignore[union-attr]
+        fake_additional_inputs = [x.meta["val"] for x in fx_additional_inputs]  # type: ignore[union-attr]
```
As these are used identically below for carried and additional inputs, would it make sense to combine them upstream to something like `all_inputs` and then do fx, fake, and use that single list below?

aakhundov commented on torch/_inductor/lowering.py:
```diff
@@ -5887,14 +5887,14 @@ def cond(pred, true_fn, false_fn, operands):
-def while_loop(cond_fn, body_fn, operands):
-    if any(map(is_triton, operands)):
+def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs):
+    if any(map(is_triton, carried_inputs)) or any(map(is_triton, additional_inputs)):
```
Nit: combine `carried_inputs + additional_inputs` instead?

aakhundov commented on torch/_library/fake_class_registry.py:
```diff
@@ -0,0 +1,234 @@
+import logging
```
Seems this file inadvertently made its way to this PR? :)

aakhundov reviewed (commented):
Thanks a lot @ydwu4 for extending the `torch.while_loop` API to handle additional inputs in a more efficient way (without need for cloning in each iteration)! Left a few comments, mostly nits.

ydwu4 committed:
Update on "[while_loop] support closures"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] support closures"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -21,31 +23,47 @@
 
 class WhileLoopOp(HigherOrderOperator):
-    def __call__(self, cond_fn, body_fn, operands):
+    def __call__(self, cond_fn, body_fn, carried_inputs, additional_inputs):
```
What are the chances you can make this positional-only lol.

e.g.:
```
def __call__(self, cond_fn, body_fn, carried_inputs, additional_inputs, /)
```

ydwu4 replied:
Yeah, I think we can do that.

I just found previously, we're not using this WhileLoopOp but the plain HigherOrderOp. Updated it to use the WhileLoopOp for checking the inputs. Will add more tests for invalide inputs as a follow-up.

zou3519 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -21,31 +23,47 @@
 
 class WhileLoopOp(HigherOrderOperator):
-    def __call__(self, cond_fn, body_fn, operands):
+    def __call__(self, cond_fn, body_fn, carried_inputs, additional_inputs):
```
can you add type annotations for these?

ydwu4 replied:
Sure!

ydwu4 committed:
Update on "[while_loop] support closures"


We add an additional_inputs arguments to the HOP while_loop and rename the operands to carried_inputs based on offline discussion with zou3519 . This allows us to support closures, parameters and buffers.

The alternative is to pass the lifted inputs directly to outputs of body_fn. But since we want the body_fn's output to not aliasing input. We'll need to copy the inputs and remove the copies later. This is a bit more work to do. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

aakhundov commented:
@ydwu4 thanks for addressing the comments. Lots of tests are failing, could you have a look?

ydwu4 committed:
Update on "[while_loop] support closures"


We add an additional_inputs arguments to the HOP while_loop and rename the operands to carried_inputs based on offline discussion with zou3519 . This allows us to support closures, parameters and buffers.

The alternative is to pass the lifted inputs directly to outputs of body_fn. But since we want the body_fn's output to not aliasing input. We'll need to copy the inputs and remove the copies later. This is a bit more work to do. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[while_loop] support closures"


We add an additional_inputs arguments to the HOP while_loop and rename the operands to carried_inputs based on offline discussion with zou3519 . This allows us to support closures, parameters and buffers.

The alternative is to pass the lifted inputs directly to outputs of body_fn. But since we want the body_fn's output to not aliasing input. We'll need to copy the inputs and remove the copies later. This is a bit more work to do. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge
